server/lib/handlers/http.py

```diff
@@ -1,11 +1,40 @@
+import urllib
+import zipfile
+
+import httpio
 import requests
 from .common import FileLikeUrlTransferHandler
 
 
+# we should probably have a better way of doing this
+SCOPE_MAP = {'pbcconsortium.isrd.isi.edu':
+                 'https://auth.globus.org/scopes/a77ee64a-fb7f-11e5-810e-8c705ad34f60/deriva_all'}
+
+
 class Http(FileLikeUrlTransferHandler):
     def __init__(self, url, transferId, itemId, psPath, user, transferManager):
         FileLikeUrlTransferHandler.__init__(self, url, transferId, itemId, psPath, user,
                                             transferManager)
+        self.extra_headers = {}
+        parsed = urllib.parse.urlparse(url)
+        if parsed.netloc in SCOPE_MAP:
+            scope = SCOPE_MAP[parsed.netloc]
+            if 'otherTokens' in user:
+                for token in user['otherTokens']:
+                    if token['scope'] == scope:
+                        self.extra_headers['Authorization'] = 'Bearer ' + token['access_token']
+
 
     def openInputStream(self):
-        return requests.get(self.url, stream=True).raw
+        parsed = urllib.parse.urlparse(self.url)
+        if parsed.path and parsed.path.endswith('.zip') and parsed.query:
+            qs = urllib.parse.parse_qs(parsed.query)
+            path = qs['path'][0]
+            fp = httpio.open(urllib.parse.urlunparse((parsed.scheme, parsed.netloc, parsed.path,
+                                                      '', '', '')), headers=self.extra_headers)
+            zf = zipfile.ZipFile(fp)
+            return zipfile.Path(zf, path).open()
```

Comment:
```suggestion
            return zipfile.Path(zf, path).open(mode="rb")
```
I looked a into other string vs bytes issue I had and it's caused by this line. Our deployment uses py3.9 and per [docs](https://docs.python.org/3/library/zipfile.html#zipfile.Path.open):

```
Path.open(mode='r', *, pwd, **)
Invoke ZipFile.open() on the current path. Allows opening for read or write, text or binary through supported modes: ‘r’, ‘w’, ‘rb’, ‘wb’. Positional and keyword arguments are passed through to io.TextIOWrapper when opened as text and ignored otherwise. pwd is the pwd parameter to ZipFile.open().

Changed in version 3.9: Added support for text and binary modes for open. Default mode is now text.
```

Reply:
OK. But then we're basically saying that 3.8 is not supported. But then the question is why. Is the existing code not working?

Reply:
Yes, the existing code is not working with our current deployment, because it's running py3.9 and it later fails with encoding error. You can make it backward compatible:

```suggestion
            try:
                return zipfile.Path(zf, path).open(mode="rb")
            except ValueError:
                return zipfile.Path(zf, path).open()
```

Reply:
True that.